### ISSUE TITLE
Design - Group User Stories

### ISSUE THREAD

**Issue by agonz97** [bobarino/cs2340gt#13]
**M4** Log In / Log Out

Together, we must form a list of **10 or less user stories** that we want to satisfy in order to accept M4 as complete. We can form this list in the comments of this Issue or over slack.

**Comment by agonz97** [comment]
User Stories (ROUGH)
- Complete Welcome Screen
  - All persons would like to be welcomed to the app by a page displaying their options. Their options should include Log In and Registration, and be represented as separate buttons.
- Complete Log In Screen
  - All persons would like to be able to log in from a separate screen. The page should include username and password fields and a button to log in.
- Complete Registration Screen
  - All persons would like to be on a separate page when they want to register an account. The registration page should include all relevant fields and a button to register.
- Working Log In Verification Service (hard-coded, if necessary)
  - A _user_, a _worker_, and _manager_  want to be able to log in using their credentials. This service should be available from the separate Log In page.
  - An _admin_ wants to be able to log into the app using admin credentials. This service should be available from the separate Log In page.
- Working Registration Service (not DB-backed yet)
  - A _user_, a _worker_, and _manager_ want to be able to register a new account for themselves. The service should be available from the page that is linked from the registration screen.
  - An _admin_ wants to be able to register a new account for anybody, including admins. This service should be available from almost anywhere within the app.
- Complete "Logged In" Screen
  - All persons need to be able to see when they have been logged in. This screen should be different from the original welcome screen, including different options.
- Complete "Registered" Screen
  - All persons want to be able to see when they have successfully registered an account. This page should be different from the welcome screen, and should be a dead end, confirmation page.
- Working Log Out Service
  - All persons should be able to Log Out from anywhere on the application. The log out service should bring all persons back to the welcome screen.

**Comment by agonz97** [comment]
COMPLETE APPLICATION DIRECTION
_All persons_ using this application should have the same power when it comes to the Welcome / Log In screens.
- The **Welcome Screen** should be the first thing seen by _all persons_ using this app, as well as the page seen by all persons after clicking the Log Out button wherever this button is available.

- The **Log In Screen** should be reachable by _all persons_ from the Welcome Screen screen only, as it is the only way to enter the application's usable pages.

- The **Logged-In State** should be reachable by _all persons_ through Log In, and represents all pages and actions available through the application following valid Log In.

- The **Registration Screen** should be reachable by _all persons_ through the Welcome Screen, and should either be a dead end or return _all persons_ back to the Welcome Screen. Any _admin_ user should be able to reach this page from anywhere on the application, and brought back to their previously working screen after successful registration.

TL;DR
This story encompasses the working navigation and direction between the general stages of the application:
- Welcome / Home Screen
- Log In Page
- Registration Page
- A 'Logged-In' State

**Comment by agonz97** [comment]
LOG IN SERVICE

_All persons_ using this application should be able to log into the application by entering proper credentials on the Log In screen. Some criteria:
- The Log In service need not ask from the person what level credentials they are logging in with. The Log In process should look the same to _all persons_.
- The Log In service must, of course, react appropriately when given both valid and invalid credentials - including relevant 'success' and 'error' prompts.
- The Log In service should be reversible in the sense of a Log Out function, reachable from anywhere on the application.

TL;DR
This story encompasses the work required to ensure that _all persons_ can log in and out from the application provided they have valid credentials only.

**Comment by agonz97** [comment]
REGISTRATION SERVICE

After M4, we should implement an actual registration service that allows _all persons_ to register a new account viable for log in.
- In terms of navigation:
  - _All persons_ should be able to access the Registration screen from the Welcome / Home screen, while
  - Only an _admin_ should see the option from anywhere on the application.
- When it comes to the actual service:
  - The registration process should fill an account with all information needed by the application later on, and should stop registration when any relevant fields are empty
  - When it comes to what power a person has, they'll be required to enter a code to confirm registration as anything above the _user_ level.
  - Prompts should be as descriptive as necessary, and the process should be generally intuitive.
- Account registration should, after M5, be persistent.
  - This posits that account information (at least log in info) must be saved somewhere
  - Closing and reopening the app shouldn't force us to create new users for demos later on, or for the final version of the app in general
  - We should look into using a sql style database to hold this data, for professionalism as well as extra credit
  - All users must be able to edit their profiles as well, which can be probably represented through a re-registration which just overrides an account's attributes.

TL;DR
This story encompasses the working Registration service that allows users to create new accounts. There are minor differences when it comes to navigating to the registration screen based on _admin_ privilege, but the actual process should be the same for _all persons_.

**Comment by agonz97** [comment]
USER WATER REPORTS

As a _user or higher_, I would like to be able to create a water report.

- When it comes to navigation:
  - The Submit Report (SR) screen should be reachable from the main page of the logged in state.
  - The SR screen should **not** be available from anywhere prior to log in.
  - *Successful report submissions should bring _all persons_ back to the main page of the logged in state.
  - *Cancelling a report submission should bring _all persons_ back to the main page of the logged in state.
- When it comes to the actual service:
  - Report submission should not be successful if not all relevant information is entered in.
  - *Successful report submission should initiate a unique prompt to the person.
  - *Cancelling report submission should initiate a unique prompt to the person.
- Finally, when it comes to data storage:
  - Data should be stored somewhere persistent
  - SQL style database is extra credit, but probably the way we should go
  - Reports also need to be viewed in a list somehow, which would be easy if stored in a DB.

TL;DR
This story describes the ability that any _user_ and person of higher account privilege has when it comes to submitting water reports. There are three things we must consider:
- Navigation within the app to reach this service
- The service itself
- Data storage and persistence

**Comment by agonz97** [comment]
WATER AVAILABILITY MAP

There must be a Water Availability screen implemented.
- _All persons_ should be able to access this screen from the main page of the logged in state.
- The screen must display a map (any API acceptable, but google maps might be best)
  - Map display should include markers at any location water has been reported
  - Clicking on these location pins should display the water report info

**Comment by agonz97** [comment]
WATER QUALITY REPORT

As a _worker_ or _manager_, I want to be able to submit a water quality report.
- When it comes to navigation:
  - This screen should only be available from the logged in state.
  - This screen should **not** be exposed to anyone with solely _user_ account privileges.
  - Successful submission and cancellation must both navigate the person back.
- When it comes to the service:
  - All required info must be entered in order for the quality report to be submitted.
  - Successful submission and cancellations should expose relevant prompts.
- When it comes to data storage:
  - Saving / submitting a report must record this report in the model
  - _Managers_ should be able to view the quality reports as a list
  - Look back at the other stories involving data storage for notes on our methods of data storage (data persistence, DBs, etc.)

TL;DR
Any person that is a _worker_ or a _manager_ should be able to create a water quality report, while a _manager_ must also be able to view all saved reports as a list. We must consider:
- Navigation to this service
- The service itself
- Data storage and persistence

**Comment by agonz97** [comment]
WATER QUALITY HISTORY GRAPH

As a _manager_, I want to view the history of water quality in the form of a graph.
- Navigation:
  - Screen should only be available from a logged in state
  - Only a _manager_ must be exposed to this screen
  - Going 'back' from this screen should return to main page
- Service:
  - Required from the _manager_ are the year and location for the water quality history
  - Data must be fetched from previously saved water quality reports
  - Data must be displayed in the form of a graph that is easy to read
